### PR TITLE
Fix Cuda atomicExchange

### PR DESCRIPTION
**Pull request by MrBurmark**
Support cuda atomicExchange in the same manner as other cuda atomics.
Add test for atomicExchange and atomicCAS.

# Summary

- This PR is a bugfix
- It does the following:
  - Fixes atomicExch\<cuda\>(double)
  - Fixes atomicExch\<hip\>(double)

### CODE REVIEW

File: test/unit/test-atomic.cpp [hunk 1]
```diff
@@ -42,6 +42,8 @@ void testAtomicFunctionBasic()
   dest[5] = (T)0;
   dest[6] = (T)N + 1;
   dest[7] = (T)0;
+  dest[8] = (T)N;
+  dest[9] = (T)0;
```
Comment (rchen20) [inline test/unit/test-atomic.cpp]: Does the size of the new and cudaMalloc above need to become 10?

Comment (MrBurmark) [reply test/unit/test-atomic.cpp]: good catch, fixed.

File: test/unit/test-atomic.cpp [hunk 2]
```diff
@@ -68,6 +72,10 @@ void testAtomicFunctionBasic()
   EXPECT_EQ((T)4, dest[5]);
   EXPECT_EQ((T)1, dest[6]);
   EXPECT_EQ((T)13, dest[7]);
+  EXPECT_LE((T)0, dest[8]);
+  EXPECT_GT((T)N, dest[8]);
+  EXPECT_LT((T)0, dest[9]);
+  EXPECT_GE((T)N, dest[9]);
```
Comment (rchen20) [inline test/unit/test-atomic.cpp]: If you have a moment, would you mind also mirroring these changes to the HIP version of the test (testAtomicFunctionBasic_gpu() lines 263-302)? Looks great, and thanks again!

Comment (MrBurmark) [reply test/unit/test-atomic.cpp]: that's a good idea.

**Comment by rchen20** [comment]
Thank you! 👍🥇

**Pull request merged**